#!/bin/bash

#### plese cite 
# Zhou et al., in review.
# Kates, H.R., Johnson, M.G., Gardner, E.M., Zerega, N.J. and Wickett, N.J., 2018. Allele phasing has minimal impact on phylogenetic reconstruction from targeted nuclear gene sequences in a case study of Artocarpus. American journal of botany, 105(3), pp.404-416.
# Please check the oringinal website https://github.com/Bean061/Putative_Paralogs_Detection and https://github.com/mossmatters/phyloscripts/tree/master/alleles_workflow

# The goal of this script is to concatenate all the supercontigs into a single file. Given a directory generated by HybPiper called prefix. (https://github.com/mossmatters/HybPiper).

## This script is modified from Mossmatters github: https://github.com/mossmatters/phyloscripts/tree/master/alleles_workflow

#### Any questions, please contact Wenbin Zhou. wzhou10@ncsu.edu



### preparation of two parameters. ###
## make sure namelist.txt is separated by \n
echo "This script is used for concatenate all the supercontigs into a single file.
All supercontigs are stored in a new generated file named supercontig.
"

echo "Please type in the following arguments in order: 
1. The direcotory HybPiper.
2. full name of namelist.
example ./Step1.sh ./HybPiper-master/[Hybpiper_result_file]/ /[full_path]/namelist.txt
"

echo "CHECKING................................"
echo "The direcotory containing the results from HybPiper: $1."
echo "The namelist is $2"

rm -rf $1/supercontig $1/exon
mkdir $1/supercontig $1/exon

while read prefix
do


### you can changed the following pathway if you are using other target enrichment software.
cat $1/${prefix}/*/${prefix}/sequences/intron/*_supercontig.fasta > $1/supercontig/${prefix}.supercontigs.fasta
cat $1/${prefix}/*/${prefix}/sequences/FNA/*.FNA > $1/exon/${prefix}.FNA.fasta
#cat $1/${prefix}/*/${prefix}/intronerate.gff > $1/exon/${prefix}.intronerate.fasta

done < $2


